honeycomb/honeycomb.go

```diff
@@ -29,83 +31,289 @@ import (
 )
 
 const (
-	defaultApiKey  = "apikey-placeholder"
 	defaultDataset = "opentelemetry"
 )
 
+// Config defines the basic configuration for the Honeycomb exporter.
 type Config struct {
-	// ApiKey is your Honeycomb authentication token, available from
-	// https://ui.honeycomb.io/account. default: apikey-placeholder
-	ApiKey string
-	// Dataset is the name of the Honeycomb dataset to which events will be
-	// sent. default: beeline-go
-	Dataset string
-	// Service Name identifies your application. While optional, setting this
-	// field is extremely valuable when you instrument multiple services. If set
-	// it will be added to all events as `service_name`
-	ServiceName string
-	// Debug will emit verbose logging to STDOUT when true. If you're having
-	// trouble getting the beeline to work, set this to true in a dev
-	// environment.
-	Debug bool
-	// APIHost is the hostname for the Honeycomb API server to which to send
-	// these events. default: https://api.honeycomb.io/
-	APIHost string
-	// UserAgent will set the user agent used by the exporter
-	UserAgent string
-	// OnError is the hook to be called when there is
-	// an error occurred when uploading the span data.
-	// If no custom hook is set, errors are logged.
-	OnError func(err error)
+	// APIKey is your Honeycomb authentication token, available from
+	// https://ui.honeycomb.io/account. This API key must have permission to
+	// send events.
+	//
+	// Don't have a Honeycomb account? Sign up at https://ui.honeycomb.io/signup.
+	APIKey string
 }
 
-// Exporter is an implementation of trace.Exporter that uploads a span to Honeycomb
+type exporterConfig struct {
+	dataset           string
+	serviceName       string
+	staticFields      map[string]interface{}
+	dynamicFields     map[string]func() interface{}
+	apiURL            string
+	userAgentAddendum string
+	output            libhoney.Output
```

Comment:
`libhoney.Output` is deprecated.  Could we use libhoney's `transmission.Sender` if we need raw control or (perhaps better) a `libhoney.Client` here instead? I haven't read through enough of this diff to understand what config here should be separate from or duplicated by configuration on a libhoney Client, but I'm curious. Static and dynamic fields, for example, exist in the libhoney client's builder; would it be better to use those instead of duplicating the field management in the exporter? (and I am just now realizing that this appearing in the diff may just be moving it from a different place rather than adding it -- and I'm fine saving the "swap out output for the newer thing" for a different diff.)

Reply:
I did see the deprecation note about `libhoney.Output`. I only adapted its existing use here, moving it around. It's being used only in test code, so the functional option to set it—`withHoneycombOutput`—is not exported from the package.

Lots of the tests in the _honeycomb_test.go_ file use the type `libhoney.MockOutput`. Is there a fake `libhoney.Client` we might be able to use instead?

Reply:
On the subject of field management, we are using `libhoney.Builder` here. We're only juggling these fields while preparing and validating the configuration, and then we call on `Builder.AddField` and `Builder.AddDynamicField` before returning the constructed exporter.

Reply:
There isn't a mock `libhoney.Client` but there is a mock `transmission.Sender`. If it's only being used for tests, I don't care so much, but maybe a `TODO` to swap it out with a `Sender` would be a nice addition.

Ahh, I missed the eventual use of the `Builder`. That seems fine. Thanks!